packages/contracts/src/contracts/current/forwarder/Forwarder.sol

```diff
@@ -0,0 +1,315 @@
+        for (uint256 i = 0; i < orders.length; i++) {
+            totalFeeAmount = safeAdd(totalFeeAmount, orders[i].takerFee);
+        }
+        if (totalFeeAmount > 0) {
+            // Fees are required for these orders. Buy enough ZRX to cover the future fill
+            Exchange.FillResults memory feeTokensResult = buyFeeTokens(feeOrders, feeSignatures, totalFeeAmount);
+            totalTakerAmountSpent = safeAdd(totalTakerAmountSpent, feeTokensResult.takerAssetFilledAmount);
+        }
+        address makerTokenAddress;
+        uint256 tokenId;
+        for (uint256 n = 0; n < orders.length; n++) {
+            Exchange.FillResults memory fillOrderResults = EXCHANGE.fillOrder(
+                orders[n],
+                orders[n].takerAssetAmount,
+                signatures[n]
+            );
+            // Fail if it wasn't fully filled otherwise we will keep WETH
+            require(fillOrderResults.takerAssetFilledAmount == orders[n].takerAssetAmount, ERROR_FAILED_TO_FILL_ALL_ORDERS);
+            totalTakerAmountSpent = safeAdd(totalTakerAmountSpent, fillOrderResults.takerAssetFilledAmount);
+            makerTokenAddress = readAddress(orders[n].makerAssetData, 1);
+            tokenId = readUint256(orders[n].makerAssetData, 21);
+            transferNFTToken(makerTokenAddress, msg.sender, tokenId);
+        }
+        // Prevent a user from over estimating the amount of ETH required
+        require(isAcceptableThreshold(msg.value, totalTakerAmountSpent), ERROR_UNACCEPTABLE_THRESHOLD);
+    }
+
+    function marketSellTokens(
+        Order[] memory orders,
+        bytes[] memory signatures,
+        Order[] memory feeOrders,
+        bytes[] memory feeSignatures,
+        uint256 sellTokenAmount)
+        private
+        returns (Exchange.FillResults memory totalFillResult)
+    {
+        uint256 takerTokenBalance = sellTokenAmount;
+        address makerTokenAddress = readAddress(orders[0].makerAssetData, 1);
+
+        Exchange.FillResults memory tokensSellQuote = marketSellOrdersQuote(orders, sellTokenAmount);
+        Exchange.FillResults memory requestedTokensResult;
+        if (tokensSellQuote.takerFeePaid > 0) {
+            // Fees are required for these orders
+            // Buy enough ZRX to cover the future market sell
+            Exchange.FillResults memory feeTokensResult = buyFeeTokens(feeOrders, feeSignatures, tokensSellQuote.takerFeePaid);
+            takerTokenBalance = safeSub(takerTokenBalance, feeTokensResult.takerAssetFilledAmount);
+            // Make our market sell to buy the requested tokens with the remaining balance
+            requestedTokensResult = EXCHANGE.marketSellOrders(orders, takerTokenBalance, signatures);
+            // It's possibile to over-buy fees by a small amount as the marketSellQuote was based on 100% of ETH
+            // and marketSell (after fee abstraction) less than 100%. If a user uses this enough it may be worth withdrawing
```

Comment:
Not sure I fully understand why it's possible of over-buy fees. Is it a rounding issue?

Reply:
The original quote was for `100ETH` and `3 ZRX` in fees, we then buy the `3 ZRX` for the `100ETH` trade, leaving us with `99.998ETH` and giving us `3ZRX`.

We can only fill the marketSell for the remaining `99.998ETH`, which will be slightly less in fees `2.99X` so there is a little over bought due to the two steps being distinct.